Cannot open compiler generated file: '"Debug\\".obj' with double backslash in the path

User @sowsem found this error when compiling with `/Fo"Debug\\" `:

```
ClCompile:
  C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\CL.exe /c /nologo /W0 /WX- /Od /Oy- /D _MBCS /Gm- /EHsc /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /Fo"Debug\\" /Fd"Debug\vc120.pdb" /Gd /TC /analyze- /errorReport:queue CMakeCCompilerId.c
  C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\clcache.py Found real compiler binary at 'C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\cl_real.exe'
  C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\clcache.py Parsing given commandline '['@C:\\Users\\davido\\AppData\\Local\\Temp\\tmpf22c581ab5514910a81a2a62e85395c1.rsp']'
  C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\clcache.py Expanded commandline '[u'/c', u'/nologo', u'/W0', u'/WX-', u'/Od', u'/Oy-', u'/D', u'_MBCS', u'/Gm-', u'/EHsc', u'/RTC1', u'/MDd', u'/GS', u'/fp:precise', u'/Zc:wchar_t', u'/Zc:forScope', u'/Fo"Debug\\\\"', u'/Fd"Debug\\vc120.pdb"', u'/Gd', u'/TC', u'/analyze-', u'/errorReport:queue', u'CMakeCCompilerId.c']'
  C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\clcache.py Compiler output file: 'Debug\\'
  C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\clcache.py Invoking real compiler as 'C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\cl_real.exe /showIncludes /c /nologo /W0 /WX- /Od /Oy- /D _MBCS /Gm- /EHsc /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /Fo"Debug\\" /Fd"Debug\vc120.pdb" /Gd /TC /analyze- /errorReport:queue CMakeCCompilerId.c'
  C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\clcache.py Real compiler returned code 1
  C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\clcache.py Finished. Exit code 1
  CMakeCCompilerId.c
C:\dev\projects\projectx-client\projects\projectx_win_tests\CMakeFiles\3.4.3\CompilerIdC\CMakeCCompilerId.c : fatal error C1083: Cannot open compiler generated file: '"Debug\\".obj': Invalid argument [C:\dev\projects\projectx-client\projects\projectx_win_tests\CMakeFiles\3.4.3\CompilerIdC\CompilerIdC.vcxproj]
```

from https://gist.github.com/sowsem/5080d43be4b32fea828a8f62cf8055ab

This revealed by a change in #105. The actual bug is that 

```
        # Strip quotes around file names; seems to happen with source files
        # with spaces in their names specified via a response file generated
        # by Visual Studio.
        if outputFile.startswith('"') and outputFile.endswith('"'):
            outputFile = outputFile[1:-1]
```

does not happen in the correct code branch in analyzeCommandLine(). So `/Fo"Debug\\"` does not become output file `Debug\CMakeCCompilerId.obj` but `"Debug\"`.

I tried this before, but the result is always:

-- Configuring BuildId for target name: WindowsVS2013-Tests
-- The C compiler identification is unknown
-- The CXX compiler identification is unknown
CMake Error in CMakeLists.txt:
  No CMAKE_C_COMPILER could be found.



CMake Error in CMakeLists.txt:
  No CMAKE_CXX_COMPILER could be found.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,8 @@ clcache changelog
  * Improvement: better protection against storing corrupt objects in the cache
    in case clcache is terminated in the middle of storing a new cache entry.
  * Bugfix: Command files with multiple lines are now handled correctly.
+ * Bugfix: Command files which contribute arguments ending in a backslash are
+   now parsed correctly (GH #108).
  * Bugfix: Properly handle non-ASCII compiler output (GH #64)
 
 ## clcache 3.1.0 (2016-06-09)

--- a/clcache.py
+++ b/clcache.py
@@ -614,21 +614,6 @@ def printTraceStatement(msg):
         print(os.path.join(scriptDir, "clcache.py") + " " + msg)
 
 
-def extractArgument(argument):
-    # If there are quotes from both sides of argument, remove them
-    # "-Isome path" must becomse -Isome path
-    if not argument:
-        return ""
-
-    if argument == '"':
-        raise Exception("Invalid argument: '\"'")
-
-    if argument[0] == '"' and argument[-1] == '"':
-        argument = argument[1:-1] # Cut first and last character
-
-    return argument.strip()
-
-
 class CommandLineTokenizer(object):
     def __init__(self, content):
         self.argv = []
@@ -704,8 +689,7 @@ class CommandLineTokenizer(object):
 
 
 def splitCommandsFile(content):
-    tokenizer = CommandLineTokenizer(content)
-    return [extractArgument(arg) for arg in tokenizer.argv]
+    return CommandLineTokenizer(content).argv
 
 
 def expandCommandLine(cmdline):

--- a/clcache.py
+++ b/clcache.py
@@ -34,6 +34,7 @@ from __future__ import print_function
 # All string literals are unicode strings. Requires Python 3.3+
 # (http://python-future.org/faq.html#which-versions-of-python-does-python-future-support)
 from __future__ import unicode_literals
+from __future__ import division
 
 from ctypes import windll, wintypes
 import codecs
@@ -692,7 +693,7 @@ class CommandLineTokenizer(object):
 
         followedByDoubleQuote = self._pos < len(self._content) and self._content[self._pos] == '"'
         if followedByDoubleQuote:
-            self._token += '\\' * (numBackslashes / 2)
+            self._token += '\\' * (numBackslashes // 2)
             if numBackslashes % 2 == 0:
                 self._pos -= 1
             else:

--- a/clcache.py
+++ b/clcache.py
@@ -628,33 +628,84 @@ def extractArgument(argument):
     return argument.strip()
 
 
-def splitCommandsLine(line):
-    # Note, we must treat lines in quotes as one argument. We do not use shlex
-    # since seems it difficult to set up it to correctly parse escaped quotes.
-    i = 0
-    wordStart = None
-    insideQuotes = False
-    result = []
-    while i < len(line):
-        if line[i] == ' ' and not insideQuotes and wordStart is not None:
-            result.append(extractArgument(line[wordStart:i]))
-            wordStart = None
-        if line[i] == '"':
-            insideQuotes = not insideQuotes
-        if line[i] != ' ' and wordStart is None:
-            wordStart = i
-        i += 1
+class CommandLineTokenizer(object):
+    def __init__(self, content):
+        self.argv = []
+        self._content = content
+        self._pos = 0
+        self._token = ''
+        self._parser = self._initialState
 
-    if wordStart is not None:
-        result.append(extractArgument(line[wordStart:]))
-    return result
+        while self._pos < len(self._content):
+            self._parser = self._parser(self._content[self._pos])
+            self._pos += 1
+
+        if self._token:
+            self.argv.append(self._token)
+
+    def _initialState(self, currentChar):
+        if currentChar.isspace():
+            return self._initialState
+
+        if currentChar == '"':
+            return self._quotedState
+
+        if currentChar == '\\':
+            self._parseBackslash()
+            return self._unquotedState
+
+        self._token += currentChar
+        return self._unquotedState
+
+    def _unquotedState(self, currentChar):
+        if currentChar.isspace():
+            self.argv.append(self._token)
+            self._token = ''
+            return self._initialState
+
+        if currentChar == '"':
+            return self._quotedState
+
+        if currentChar == '\\':
+            self._parseBackslash()
+            return self._unquotedState
+
+        self._token += currentChar
+        return self._unquotedState
+
+    def _quotedState(self, currentChar):
+        if currentChar == '"':
+            return self._unquotedState
+
+        if currentChar == '\\':
+            self._parseBackslash()
+            return self._quotedState
+
+        self._token += currentChar
+        return self._quotedState
+
+    def _parseBackslash(self):
+        numBackslashes = 0
+        while self._pos < len(self._content) and self._content[self._pos] == '\\':
+            self._pos += 1
+            numBackslashes += 1
+
+        followedByDoubleQuote = self._pos < len(self._content) and self._content[self._pos] == '"'
+        if followedByDoubleQuote:
+            self._token += '\\' * (numBackslashes / 2)
+            if numBackslashes % 2 == 0:
+                self._pos -= 1
+            else:
+                self._token += '"'
+        else:
+            self._token += '\\' * numBackslashes
+            self._pos -= 1
+
 
 def splitCommandsFile(content):
-    ret = []
-    for line in content.splitlines(True):
-        ret.extend(splitCommandsLine(line.strip()))
+    tokenizer = CommandLineTokenizer(content)
+    return [extractArgument(arg) for arg in tokenizer.argv]
 
-    return ret
 
 def expandCommandLine(cmdline):
     ret = []

--- a/unittests.py
+++ b/unittests.py
@@ -88,6 +88,13 @@ class TestSplitCommandsFile(BaseTest):
         self._genericTest('   -A -B -C', ['-A', '-B', '-C'])
         self._genericTest('-A -B -C   ', ['-A', '-B', '-C'])
 
+    def testMicrosoftExamples(self):
+        # https://msdn.microsoft.com/en-us/library/17w5ykft.aspx
+        self._genericTest(r'"abc" d e', ['abc', 'd', 'e'])
+        self._genericTest(r'a\\b d"e f"g h', [r'a\\b', 'de fg', 'h'])
+        self._genericTest(r'a\\\"b c d', [r'a\"b', 'c', 'd'])
+        self._genericTest(r'a\\\\"b c" d e', [r'a\\b c', 'd', 'e'])
+
     def testQuotesAroundArgument(self):
         self._genericTest(r'/Fo"C:\out dir\main.obj"', [r'/Fo"C:\out dir\main.obj"'])
         self._genericTest(r'/c /Fo"C:\out dir\main.obj"', ['/c', r'/Fo"C:\out dir\main.obj"'])

--- a/unittests.py
+++ b/unittests.py
@@ -108,10 +108,17 @@ class TestSplitCommandsFile(BaseTest):
         self._genericTest(r'/c "/Fo"something\main.obj"" /nologo', ['/c', r'/Fosomething\main.obj', '/nologo'])
 
     def testBackslashBeforeQuote(self):
+        # Pathological cases of escaping the quote incorrectly.
         self._genericTest(r'/Fo"C:\out dir\"', [r'/FoC:\out dir"'])
         self._genericTest(r'/c /Fo"C:\out dir\"', ['/c', r'/FoC:\out dir"'])
         self._genericTest(r'/Fo"C:\out dir\" /nologo', [r'/FoC:\out dir" /nologo'])
         self._genericTest(r'/c /Fo"C:\out dir\" /nologo', ['/c', r'/FoC:\out dir" /nologo'])
+
+        # Sane cases of escaping the backslash correctly.
+        self._genericTest(r'/Fo"C:\out dir\\"', [r'/FoC:\out dir' '\\'])
+        self._genericTest(r'/c /Fo"C:\out dir\\"', ['/c', r'/FoC:\out dir' '\\'])
+        self._genericTest(r'/Fo"C:\out dir\\" /nologo', [r'/FoC:\out dir' '\\', r'/nologo'])
+        self._genericTest(r'/c /Fo"C:\out dir\\" /nologo', ['/c', r'/FoC:\out dir' '\\', r'/nologo'])
 
     def testVyachselavCase(self):
         self._genericTest(

--- a/unittests.py
+++ b/unittests.py
@@ -47,30 +47,6 @@ class BaseTest(unittest.TestCase):
         super(BaseTest, self).__init__(*args, **kwargs)
 
 
-class TestExtractArgument(BaseTest):
-    def testSimple(self):
-        # Keep
-        self.assertEqual(clcache.extractArgument(r''), r'')
-        self.assertEqual(clcache.extractArgument(r'1'), r'1')
-        self.assertEqual(clcache.extractArgument(r'myfile.cpp'), r'myfile.cpp')
-        self.assertEqual(
-            clcache.extractArgument(r'/DEXTERNAL_DLL=__declspec(dllexport)'),
-            r'/DEXTERNAL_DLL=__declspec(dllexport)')
-        self.assertEqual(clcache.extractArgument(r'-DVERSION=\\"1.0\\"'), r'-DVERSION=\\"1.0\\"')
-        self.assertEqual(clcache.extractArgument(r'-I"..\.."'), r'-I"..\.."')
-
-        # Extract
-        self.assertEqual(
-            clcache.extractArgument(r'"-IC:\Program Files\Lib1"'),
-            r'-IC:\Program Files\Lib1')
-        self.assertEqual(
-            clcache.extractArgument(r'"/Fo"CrashReport.dir\Release\""'),
-            r'/Fo"CrashReport.dir\Release\"')
-        self.assertEqual(
-            clcache.extractArgument(r'"-DWEBRTC_SVNREVISION=\"Unavailable(issue687)\""'),
-            r'-DWEBRTC_SVNREVISION=\"Unavailable(issue687)\"')
-
-
 class TestSplitCommandsFile(BaseTest):
     def _genericTest(self, commandLine, expected):
         self.assertEqual(clcache.splitCommandsFile(commandLine), expected)

--- a/unittests.py
+++ b/unittests.py
@@ -108,10 +108,10 @@ class TestSplitCommandsFile(BaseTest):
         self._genericTest(r'/c "/Fo"something\main.obj"" /nologo', ['/c', r'/Fosomething\main.obj', '/nologo'])
 
     def testBackslashBeforeQuote(self):
-        self._genericTest(r'/Fo"C:\out dir\"', [r'/Fo"C:\out dir\"'])
-        self._genericTest(r'/c /Fo"C:\out dir\"', ['/c', r'/Fo"C:\out dir\"'])
-        self._genericTest(r'/Fo"C:\out dir\" /nologo', [r'/Fo"C:\out dir\"', '/nologo'])
-        self._genericTest(r'/c /Fo"C:\out dir\" /nologo', ['/c', r'/Fo"C:\out dir\"', '/nologo'])
+        self._genericTest(r'/Fo"C:\out dir\"', [r'/FoC:\out dir"'])
+        self._genericTest(r'/c /Fo"C:\out dir\"', ['/c', r'/FoC:\out dir"'])
+        self._genericTest(r'/Fo"C:\out dir\" /nologo', [r'/FoC:\out dir" /nologo'])
+        self._genericTest(r'/c /Fo"C:\out dir\" /nologo', ['/c', r'/FoC:\out dir" /nologo'])
 
     def testVyachselavCase(self):
         self._genericTest(
@@ -119,10 +119,10 @@ class TestSplitCommandsFile(BaseTest):
             [
                 r'-IC:\Program files\Some library',
                 r'-DX=1',
-                r'-DVERSION=\"1.0\"',
+                r'-DVERSION="1.0"',
                 r'-I..\..',
-                r'-I"..\..\lib"',
-                r'-DMYPATH=\"C:\Path\"'
+                r'-I..\..\lib',
+                r'-DMYPATH="C:\Path"'
             ])
 
     def testLineEndings(self):

--- a/unittests.py
+++ b/unittests.py
@@ -96,16 +96,16 @@ class TestSplitCommandsFile(BaseTest):
         self._genericTest(r'a\\\\"b c" d e', [r'a\\b c', 'd', 'e'])
 
     def testQuotesAroundArgument(self):
-        self._genericTest(r'/Fo"C:\out dir\main.obj"', [r'/Fo"C:\out dir\main.obj"'])
-        self._genericTest(r'/c /Fo"C:\out dir\main.obj"', ['/c', r'/Fo"C:\out dir\main.obj"'])
-        self._genericTest(r'/Fo"C:\out dir\main.obj" /nologo', [r'/Fo"C:\out dir\main.obj"', '/nologo'])
-        self._genericTest(r'/c /Fo"C:\out dir\main.obj" /nologo', ['/c', r'/Fo"C:\out dir\main.obj"', '/nologo'])
+        self._genericTest(r'/Fo"C:\out dir\main.obj"', [r'/FoC:\out dir\main.obj'])
+        self._genericTest(r'/c /Fo"C:\out dir\main.obj"', ['/c', r'/FoC:\out dir\main.obj'])
+        self._genericTest(r'/Fo"C:\out dir\main.obj" /nologo', [r'/FoC:\out dir\main.obj', '/nologo'])
+        self._genericTest(r'/c /Fo"C:\out dir\main.obj" /nologo', ['/c', r'/FoC:\out dir\main.obj', '/nologo'])
 
     def testDoubleQuoted(self):
-        self._genericTest(r'"/Fo"something\main.obj""', [r'/Fo"something\main.obj"'])
-        self._genericTest(r'/c "/Fo"something\main.obj""', ['/c', r'/Fo"something\main.obj"'])
-        self._genericTest(r'"/Fo"something\main.obj"" /nologo', [r'/Fo"something\main.obj"', '/nologo'])
-        self._genericTest(r'/c "/Fo"something\main.obj"" /nologo', ['/c', r'/Fo"something\main.obj"', '/nologo'])
+        self._genericTest(r'"/Fo"something\main.obj""', [r'/Fosomething\main.obj'])
+        self._genericTest(r'/c "/Fo"something\main.obj""', ['/c', r'/Fosomething\main.obj'])
+        self._genericTest(r'"/Fo"something\main.obj"" /nologo', [r'/Fosomething\main.obj', '/nologo'])
+        self._genericTest(r'/c "/Fo"something\main.obj"" /nologo', ['/c', r'/Fosomething\main.obj', '/nologo'])
 
     def testBackslashBeforeQuote(self):
         self._genericTest(r'/Fo"C:\out dir\"', [r'/Fo"C:\out dir\"'])

